### DISCUSSION TITLE
Glitch in nanochat/optim.py?

### DISCUSSION THREAD

**Discussion by JohnDrakopoulos** [karpathy/nanochat#503] [General]
It seems the function `_reduce_adamw `in _nanochat/optim.py_ will miss updating some of the gradients of parameters that have large size (>1024) and their size is not a multiple of world_size. In that case `grad.shape[0] % world_size` gradient sub-arrays will not be updated (in the code path within the else statement). The network can still be trained and seem to have good results but the corresponding parameters will be incorrect.

    def _reduce_adamw(self, group: dict, world_size: int) -> dict:
        """Launch async reduce ops for AdamW group. Returns info dict with per-param infos."""
        param_infos = {}
        for p in group['params']:
            grad = p.grad
            if p.numel() < 1024:
                # Small params: all_reduce (no scatter/gather needed)
                future = dist.all_reduce(grad, op=dist.ReduceOp.AVG, async_op=True).get_future()
                param_infos[p] = dict(future=future, grad_slice=grad, is_small=True)
            else:
                # Large params: reduce_scatter
                rank_size = grad.shape[0] // world_size
                grad_slice = torch.empty_like(grad[:rank_size])
                future = dist.reduce_scatter_tensor(grad_slice, grad, op=dist.ReduceOp.AVG, async_op=True).get_future()
                param_infos[p] = dict(future=future, grad_slice=grad_slice, is_small=False)
        return dict(param_infos=param_infos)

**Comment by karpathy** [comment]
This doesn't happen in practice right now but could happen in principle. I had an assert for this earlier but somehow it got lost in a refactor. It's only mentioned in a docstring that this must be divisible. Brought back the assert. https://github.com/karpathy/nanochat/commit/98eed6df189e395056c34621043d082878df392f

**Reply by JohnDrakopoulos** [reply]
Alternatively, we can check for divisibility in advance and remove the assert. However, this is a tiny optimization. But I agree that we need a check somewhere.

**Reply by JohnDrakopoulos** [reply]
BTW, the refactoring is a clear improvement. If you ever change optim.py again, you may want to replace the following two lines (lines 470, 471):
```
            owned_params = [params[start_idx + i] for i in range(num_owned)]
            stacked_owned = torch.stack(owned_params)
```
with
`            stacked_owned = torch.stack(params[start_idx: start_idx + num_owned])`
(For these kind of slices, list comprehension is about 10 times faster.)
